File: o2/Exercise2.c
#include <stdlib.h>
#include <stdio.h>
#include <time.h>

double recursive_v1(double base, int exponent){
	if (exponent == 1){
		return base;
	}
	else {
		return base * recursive_v1(base, exponent - 1);
	}
}

double recursive_v2(double base, int exponent){
	if (exponent == 1){
		return base;
	}
	else if ((exponent & 1) == 1){
		double temp = recursive_v2(base * base, (exponent - 1) / 2);
		return base * (temp);
	}
	else {
		double temp = recursive_v2(base * base, exponent / 2);
		return temp;
	}
}

int main(){
	
	printf("Jeg elsker Irja!! <3 \n");

	double base = 2.0;

	int exponents[] = {128, 256, 512, 1024, 2048, 4096, 8192, 16384};
	double result = 0;


	long start, end;
	
	for (int i = 0; i < 8; i++){
		
		start = clock();

		result = recursive_v1(base, exponents[i]);

		
		end = clock();
		
		double timeTaken = (end - start) / CLOCKS_PER_SEC;

		printf("Exponent number %f ^ %d = %f\n. Time taken: %f\n", base, exponents[i], result, timeTaken);
	}

	return 0;
}

File: o2/src/Exercise2.java
package o2.src;

import java.lang.Math;
public class Exercise2 {
    
    public static void main(String[] args) {
        double base = 1.0002;

        int[] exponents = new int[] {10, 100, 1000, 10000};
        
        long start;
        double result;
        double duration;


        for (int exponent : exponents) {
            start = System.nanoTime();
            result = recursive_exponent_v1(base, exponent);
            duration = (System.nanoTime() - start);
            System.out.println("Time elapsed in recursive_exponent_v1(), " + result + " is: " + duration + " ns\n");
        }
        for (int exponent : exponents) {
            start = System.nanoTime();
            result = recursive_exponent_v2(base, exponent);
            duration = (System.nanoTime() - start);
            System.out.println("Time elapsed in recursive_exponent_v2(), " + result + " is: " + duration + " ns\n");
        }
        for (int exponent : exponents) {
            start = System.nanoTime();
            result = Math.pow(base, exponent);
            duration = (System.nanoTime() - start);
            System.out.println("Time elapsed in Math.pow(), " + result + " is: " + duration + " ns\n");
        }
    }

    public static double recursive_exponent_v1(double base, int exponent){
        if (exponent == 1){
            return base;
        }
        else {
            return base * recursive_exponent_v1(base, exponent - 1);
        }
    }


    public static double recursive_exponent_v2(double base, int exponent) {
        if (exponent == 1){
            return base;
        }
        else if ((exponent & 1) == 1) {
            double temp = recursive_exponent_v2(base * base, (exponent - 1) / 2);
            return base * temp; 
        } else {
            double temp = recursive_exponent_v2(base * base, exponent / 2);
            return temp;
        }
    }

}

